modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxUnlockManager.java

```diff
@@ -0,0 +1,321 @@
+import org.apache.ignite.internal.tx.message.LockReleaseMessage;
+import org.apache.ignite.internal.tx.message.TxMessageGroup;
+import org.apache.ignite.internal.tx.message.TxMessagesFactory;
+import org.apache.ignite.network.ClusterService;
+import org.apache.ignite.network.NetworkMessage;
+import org.apache.ignite.tx.TransactionException;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Releases TX locks.
+ */
+public class TxUnlockManager implements IgniteComponent {
+    /** The logger. */
+    private static final IgniteLogger LOG = Loggers.forClass(TxUnlockManager.class);
+
+    /** Network timeout. */
+    private static final long RPC_TIMEOUT = 3000;
+    /** Tx messages factory. */
+    private static final TxMessagesFactory FACTORY = new TxMessagesFactory();
+
+    private static final int AWAIT_PRIMARY_REPLICA_TIMEOUT = 10;
+
+    private static final int ATTEMPTS_TO_CLEANUP_REPLICA = 5;
+
+    /** Cluster service. */
+    private final ClusterService clusterService;
+    /** Transaction manager. */
+    private final TxManager txManager;
+    /** Lock manager. */
+    private final LockManager lockManager;
+    /**
+     * Placement driver.
+     */
+    private final PlacementDriver placementDriver;
+    /** Hybrid clock. */
+    private final HybridClock hybridClock;
+
+    /**
+     * The constructor.
+     *
+     * @param clusterService Cluster service.
+     * @param txManager Transaction manager.
+     * @param lockManager Lock manager.
+     * @param placementDriver Placement driver.
+     * @param clock A hybrid logical clock.
+     */
+    public TxUnlockManager(
+            ClusterService clusterService,
+            TxManager txManager,
+            LockManager lockManager,
+            PlacementDriver placementDriver,
+            HybridClock clock
+    ) {
+        this.clusterService = clusterService;
+        this.txManager = txManager;
+        this.lockManager = lockManager;
+        this.placementDriver = placementDriver;
+        this.hybridClock = clock;
+    }
+
+    @Override
+    public void start() {
+        clusterService.messagingService().addMessageHandler(TxMessageGroup.class, (msg, sender, correlationId) -> {
+            if (msg instanceof LockReleaseMessage) {
+                processLockRelease((LockReleaseMessage) msg, sender, correlationId);
+            }
+        });
+    }
+
+    /**
+     * Sends unlock request to the primary nodes of each one of {@code partitions}.
+     *
+     * @param partitions Enlisted partition groups.
+     * @param commit {@code true} if a commit requested.
+     * @param commitTimestamp Commit timestamp ({@code null} if it's an abort).
+     * @param txId Transaction id.
+     * @return Completable future of Void.
+     */
+    public CompletableFuture<Void> unlock(
+            Collection<TablePartitionId> partitions,
+            boolean commit,
+            @Nullable HybridTimestamp commitTimestamp,
+            UUID txId
+    ) {
```

Comment:
How would we act here in the case of recovery if the partition collection is empty? We still have a node that sent the `InitiateRecovery` request.